worker/credentialvalidator/worker.go

```diff
@@ -64,42 +70,42 @@ func NewWorker(config Config) (worker.Worker, error) {
 		return nil, errors.Trace(err)
 	}
 
-	// It is possible that this model is on a cloud that does not
-	// necessarily require authentication. Consequently, a model
-	// is created without a set credential. However, later on,
-	// a credential may be added to a model. This worker will need to
-	// be restarted as there will be no credential watcher set to react to
-	// changes.
-	// TODO (anastasiamac 2018-05-30) when model-credential relationship no
-	// longer resides on model itself but has a dedicated document in mongo,
-	// we can have a watcher that will support the above corner case.
+	// This worker needs to monitor both the changes to the credential content that
+	// this model uses as well as what credential the model uses.
+	// It needs to be restarted if there is a change in either.
+	mcw, err := config.Facade.WatchModelCredential()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
 
 	v := &validator{
-		validatorFacade: config.Facade,
-		credential:      mc,
+		validatorFacade:        config.Facade,
+		credential:             mc,
+		modelCredentialWatcher: mcw,
 	}
 
+	// The watcher needs to be added to the worker's catacomb plan
+	// here in order to be controlled by this worker's lifecycle events:
+	// for example, to be destroyed when this worker is destroyed, etc.
+	// We also add the watcher to the Plan.Init collection to ensure that
+	// the worker's Plan.Work method is executed after the watcher
+	// is initialised and watcher's changes collection obtains the changes.
+	// Watchers that are added using catacomb.Add method
```

Comment:
I don't think this comment's right - the watcher's changes collection is initialised when it's created. Using `catacomb.Add` should be the same as putting the worker in `plan.Init` (they both use `catacomb.add` under the hood) - the only difference is that if the catacomb's been killed for some reason before the watcher is `.Add`ed the watcher will be stopped immediately, which isn't needed if the watcher is in `plan.Init` (since the catacomb couldn't have been killed).

It's fine to put the watcher in `plan.Init`, but if there's some subtle bug with adding watchers via `catacomb.Add` then we have a big problem because lots of workers do that.

Reply:
So, the comment was made with relation to the actual implementation of catacomb.Invoke().
I have done a lot of investigations and testing when I first wrote this worker (and this comment... note it was just moved not really added). See
https://github.com/juju/worker/blob/v1/catacomb/catacomb.go#L91

I needed the watchers in catacomb by the time I invoked to react correctly to initial change. I think maybe majority of the situations you describe do not care for initial state of events and only care about subsequent changes?...I don't know the answer for the situations you are thinking of, but I do know that here I needed watchers in Plan.Init.